cidc_api/resources/upload_jobs.py

```diff
@@ -86,18 +93,86 @@ def get_upload_job(upload_job: UploadJobs):
     return upload_job
 
 
+def requires_upload_token_auth(endpoint):
+    """
+    Decorator that adds "upload token" authentication to an endpoint. 
+    The provided endpoint must include the upload job id as a URL param, i.e., 
+    `<int:upload_job>`. This upload job ID is used to look up the relevant upload_job
+    and check its `token` field against the user-provided `token` query parameter.
+    If authentication and upload job record lookup succeeds, pass the upload job record
+    in the `upload_job` kwarg to `endpoint`.
+    """
+    # Flag this endpoint as authenticated
+    endpoint.is_protected = True
+
+    token_schema = Schema.from_dict({"token": fields.Str(required=True)})(
+        # Don't throw an error if there are unknown query params in addition to "token"
+        unknown=INCLUDE
+    )
+
+    @wraps(endpoint)
+    @use_args(token_schema, location="query")
+    def wrapped(args, *pos_args, **kwargs):
+        # Attempt identity token authentication to get user info
+        user = _get_user()
+
+        # Try to find the upload_job associated with this request
+        upload_job = _get_upload_job(user, kwargs["upload_job"])
+
+        # Check that the user-provided upload token matches the saved upload token
+        token = args["token"]
+        if str(token) != str(upload_job.token):
+            raise Unauthorized("upload_job token authentication failed")
+
+        # Pass the looked-up upload_job record to `endpoint` via the `upload_job` keyword argument
+        kwargs["upload_job"] = upload_job
+
+        return endpoint(*pos_args, **kwargs)
+
+    def _get_user():
```

Comment:
nit: not sure why you made this and `_get_upload_job` as part of decorator definition and not module level declaration